### ISSUE TITLE
Monoid instances for primitives and String can easily lead to bugs

### ISSUE THREAD

**Issue by cvogt** [scalaz/scalaz#1220]
We definitely had bugs in real code where the Monoid[String] kicked in by accident and we ended up with concatenated string. Turned out we appaneded a Map[Foo,String], when the code assumed a Map[Foo,List[String]].

Seems like Monoids of containers are less likely used wrong by accident than Monoid's of primitives. Of course that's subjective. Monoid[Char] seems too easy to mis-use too, so do others.

![asdf](https://pbs.twimg.com/media/CnXJakCWYAAwbLW.jpg)

So should we do something about that? For comparison: Monoid[Boolean] doesn't exist. Instead Monoid[Boolean @@ Conjunction]  and Monoid[Boolean @@ Disjunction] exist. That's one way to do. It not that great in the context types that occur in containers whose element types you can't easily change, like case classes, because you may not be able to tag the types.

Another alternative would be not importing the Monoids for primitives by default and only letting people selectively import the ones they want in a local scope. I know this approach is often frowned up-on and global, unique Monoid instance have strong proponents. But it is one option to consider.

What do other people think about the topic?

**Comment by aloiscochard** [comment]
Hi Chris!

I think this is too subjective to be formalized, I do not think having a sort strategy using imports is good: we want typeclass coherence, if an instance exists it should be available (in scalaz8 the instance for primitive type are declared directly in the companion of the class, so there will be available anyways).

To be honest, I don't really see a problem here, if you want to be really safe in your company code base, instead of using `Monoid[A]`, you can use `Monoid[Foo[A]]`, where `Foo` is a newtype for which you can defined precisely the instance you want and avoid others.

Coming back to your original problem: "Turned out we appaneded a Map[Foo,String], when the code assumed a Map[Foo,List[String]].".

How can that happen, shouldn't you have been explicit about `List` at some point? I would love to see the complete example if you can!

Thanks mate

**Comment by cvogt** [comment]
So for Monoid[Char] (as most people seems to agree that's a weird one), the only time it would be used in the code I personally write would be by accident as a bug. Meaning having it available at all times means a correctness risk for me. Trying to make my code safe I am of course not comfortable with that. I guess it is a trade-off how important type-class coherence is even for very rare use cases. To me it seems having those always available reduces type-safety, but I know many people have very strong opinions in favor of type-class coherence for everything.

How can it happen? Local type inference. In our business code we have lots of contextual data and need to narrow it down for each specific use case meaning usually a sequence of transformation operations including things like groupBy of nested collections, Monoids, etc. And sometimes you get those wrong and the intermediate types are slightly different. Most of the time the type system catches that. Sometimes not and this was one case. Don't have the code for the String Map case anymore, it's been a while back and the bug has been fixed.

**Comment by aloiscochard** [comment]
I'm afraid I genuinely don't understand why `Monoid[Char]` could be fundamentally seen as a weirder than `Monoid[Int]` or `Show[Char]`, and you are the first one I hear telling me that this instance is weird.

It seems from where I stand, that your opinion is biased to your very specific experience and the context of usage.

What should we do if tomorrow someone comes in and say "In my context, that instance of Monoid[Int] does not makes any sense and create bug, because I want to use only Monoid for pickles only!".

Why treating this one specifically? to be clear, I would be totally fine with having a default `Monoid[Boolean]` where we actually flip a coin to choose which one should be the default, this is just a convention for me. If it can be intuitive of course it must be, but I'm totally fine having default instance if there is multiple possible implementation. I don't think this really relate here as there is only one instance but I just wanted to make that clear.

I do really care about global coherence, and I really strongly disagree with that sentence "No the final answer either as it would treat all occurences of the same type the same way in a syntax scope, not necessarily what one wants." .... I also think it's a shared feeling among all active scalaz contributors, so it seems really hard and counter productive to push the library in the other direction.

I'm still puzzled about how it could create bug in conjunction with local inference, I mean if you got bug due to that, you might as well get the exact same bug with any other typeclass or types. So I'm really interested in a concrete example.

In other words, I mean that I suppose you could take any example your issue around `Monoid[List[Int]]` and `Monoid[Int]`, and make it equally reproducible using `Monoid[Map[A, List[A]]` and `Monid[List[A]]`... is that a good reason to condemn one of the later? That sounds totally arbitrary to me.

Also, is there anything wrong about using newtypes to get the extra safety you are looking for?

**Comment by cvogt** [comment]
>>No the final answer either as it would treat all occurences of the same type the same way in a syntax scope, not necessarily what one wants.

referred to local imports of instances not being the final answer, not to type-class coherence. I know for many people it is the final answer.

I am not concretely proposing a change by the way, just wanted to point out my problems with the status quo and find out what other people think about it.

The difference to me seems to be that if you mess up something with containers it seems to be usually caught later by the type system. When you end in primitives it does happen less often. Maybe that's too subjective of an experience. I can expand on why over chat.

We tried new-types for primitives, but they seem impractical for wrapping every single one and actually seem to not provide additional type-safety on their own when used universally because "casting" becomes so common that it requires much care to get right.

Happy to continue the discussion over chat.

My indication that people seem to tend to agree on Monoid[Char] being weird are these Twitter/IRC responses:

![image](https://cloud.githubusercontent.com/assets/274947/16876747/25ff6278-4a73-11e6-9cef-85cf980a3527.png)
![image](https://cloud.githubusercontent.com/assets/274947/16876797/673114c6-4a73-11e6-8256-904125d713ec.png)

![image](https://cloud.githubusercontent.com/assets/274947/16876779/4fa0d3f0-4a73-11e6-9bbe-edaca4515abf.png)
![image](https://cloud.githubusercontent.com/assets/274947/16876789/5d1e30c2-4a73-11e6-9d77-9484a54127ad.png)

**Comment by aloiscochard** [comment]
@cvogt okay thanks, that surely help a lot understanding where you come from, your context and your point of view.

I did reach out aarvar on IRC, and a few other folks gave me their rational and I think I agree about "we shouldn't have `Monoid[Char]`, but the reason I agree with this (and it's the same reason for all person I did ask too), it's because it is wrong to treat a `Char` as a `Int`, and that's exactly what the `Monoid` instance does... this is stupid, if you want that semantic,one should convert the `Char` to an `Int`.

But it doesn't mean we don't want an instance for it **because it is a primitive**, this is very different, and IMO too subjective to be formalized.

For example, I like `Monoid[Unit]` and use it regularly. It does not seems that today we have a general solution/approach for the issue you describe, but it is surely an interesting one... probably more typeclass limitation more than special treatment of primitives (the problem feel more general than that to me, but need more thought).

**Comment by pedrofurla** [comment]
" it's because it is wrong to treat a Char as a Int, and that's exactly what the Monoid instance does... this is stupid" 👍  very much

**Comment by vmarquez** [comment]
I don't see what it should be treated as, if not an Int?  Ultimately if you do  
`'a' + 'b'` in scala, it is treated as an Int.  Therefore, the *most reasonable* thing for the Monoid implementation is to treat it the same.  

I don't agree that this is problematic or a bug.

**Comment by jbgi** [comment]
The [javadoc](http://docs.oracle.com/javase/8/docs/api/java/lang/Character.html#unicode) mentions
> A char value, therefore, represents Basic Multilingual Plane (BMP) code points [...]

but there is [unallocated ranges](https://en.wikipedia.org/wiki/Plane_(Unicode)#Basic_Multilingual_Plane). So `'a'+'b'` is not guaranteed to be a valid code point of the BMP.

If you treat char as having no relation with unicode characters (ie. only a shorter int) then there is no problem in having a Monoid for them.
But I think the unicode interpretation for Char is more prevalent among programmers (and is the intended interpretation).

**Comment by bmjames** [comment]
@vmarquez: if you _must_ provide a `Monoid[Char]`, then that is probably the most reasonable thing for it to do. But I don't accept the premise. It shouldn't be the mission of Scalaz to provide every possible instance, and I'm not convinced that this instance deserves a place in the canon.
